File: homework/ba-nguyen/lesson-8/two-sum-iv-input-is-a-bst.ts
function findTarget(root: TreeNode | null, k: number): boolean {
  const array: number[] = [];
  inorderTraversal(root, array);

  let lo = 0;
  let hi = array.length - 1;

  while (lo < hi) {
    const sum = array[lo] + array[hi];
    if (sum === k) return true;
    if (sum < k) {
      lo++;
    } else {
      hi--;
    }
  }

  return false;
}

function inorderTraversal(node: TreeNode | null, array: number[]) {
  if (!node) return;

  inorderTraversal(node.left, array);
  array.push(node.val);
  inorderTraversal(node.right, array);
}

File: homework/Long_Nguyen/lesson_8/TwoSumIVInputIsABst.java
import java.util.ArrayList;
import java.util.List;

/**
 * Created by Long Nguyen
 */
public class TwoSumIVInputIsABst {
    public void inOrder(TreeNode root, List<Integer> cache) {
        if (root == null)
            return;
        inOrder(root.left, cache);
        cache.add(root.val);
        inOrder(root.right, cache);
    }

    public boolean findTarget(TreeNode root, int k) {
        if (root == null)
            return false;

        List<Integer> cache = new ArrayList<>();
        inOrder(root, cache);

        int first = 0;
        int second = cache.size() - 1;

        while (first < second) {
            int sum = cache.get(first) + cache.get(second);
            if (sum == k)
                return true;
            if (sum < k) {
                first++;
            } else {
                second--;
            }
        }

        return false;
    }
}
